keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,722 @@
+with a leaf template can also be a stage. For example, the following example contains 
+four stages(A, B, C, D):
+
+``` yaml
+#     A
+#    / \
+#   B   C
+#    \ /
+#     D
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: dag-diamond-
+spec:
+  entrypoint: diamond
+  templates:
+  - name: echo
+    inputs:
+      parameters:
+      - name: message
+    container:
+      image: alpine:3.7
+      command: [echo, "{{inputs.parameters.message}}"]
+  - name: diamond
+    dag:
+      tasks:
+      - name: A
+        template: echo
+        arguments:
+          parameters: [{name: message, value: A}]
+      - name: B
+        dependencies: [A]
+        template: echo
+        arguments:
+          parameters: [{name: message, value: B}]
+      - name: C
+        dependencies: [A]
+        template: echo
+        arguments:
+          parameters: [{name: message, value: C}]
+      - name: D
+        dependencies: [B, C]
+        template: echo
+        arguments:
+          parameters: [{name: message, value: D}]
+```
+
+### Workflow as An Unit
+
+#### How to suspend the whole workflow
+
+Given the diverse resource, node affinity, and toleration requirements among workflow pods, 
+determining the necessary resources for each flavor becomes challenging for the controller. 
+Users must specify workflow resources via annotations like kueue.k8s.io/max-resources, tolerations 
+with kueue.k8s.io/toleration, and node selectors with kueue.k8s.io/node-selector.
+
+In this way, only one workload will be create for one workflow, so workflows can be integrated into 
+Kueue with current job framework. Only need to implement for the specific workflow types.
+
+#### Drawback and Limitations
+
+- Inability to set distinct nodeSelectors and tolerations for multiple pod sets within a workflow.
+- Hard to inject any node selectors that Kueue could calculate from flavor assignment.
+
+#### Advantages
+
+- Simplified architecture facilitating straightforward implementation. Workload managers can 
+follow the same api to queue their workflows.
+
+### Stage as An Unit
+
+#### Workflow Framework in Kueue
+
+When we regard stage as an unit, multiple workloads will be created for one workflow. So the current
+job framework is not suitable for this case. We need to create a new `workflow framework` for the 
+workflows. The generic workflow framework reconcile workflows in following way:
+
+``` go
+type GenericWorkflow interface {
+  GetActiveStages() []GenericJob
+}
+
+
+func (r *workflowReconciler)ReconcileGenericWorkflow(ctx context.Context, req ctrl.Request, wf GenericWorkflow) {
+  dropFinalizers := GetWorkflowInstance()
+  if dropFinalizers {
+    DropFinalizers()
+    return 
+  }
+
+  // currently we do not support parent-workload for workflow
+  // stage should be implemented as GenericJob
+  stages := wf.GetActiveStages()
+  for _, stage := range stages {
+    // 1. make sure there is only a single existing instance of the workload.
```

Comment:
What do you mean by single existing instance of workload here ?

Reply:
For a specific stage which may refer to a stageGroup in workflow's status in workflow, we create a kueue.workload to track the resource requests.